c_glib/arrow-dataset-glib/scanner.cpp

```diff
@@ -0,0 +1,524 @@
+                                    gad_scan_task,
+                                    G_TYPE_OBJECT)
+
+#define GAD_SCAN_TASK_GET_PRIVATE(obj)         \
+  static_cast<GADScanTaskPrivate *>(           \
+     gad_scan_task_get_instance_private(       \
+       GAD_SCAN_TASK(obj)))
+
+static void
+gad_scan_task_finalize(GObject *object)
+{
+  auto priv = GAD_SCAN_TASK_GET_PRIVATE(object);
+
+  priv->scan_task.~shared_ptr();
+
+  G_OBJECT_CLASS(gad_scan_task_parent_class)->finalize(object);
+}
+
+static void
+gad_scan_task_set_property(GObject *object,
+                           guint prop_id,
+                           const GValue *value,
+                           GParamSpec *pspec)
+{
+  auto priv = GAD_SCAN_TASK_GET_PRIVATE(object);
+
+  switch (prop_id) {
+  case PROP_SCAN_TASK:
+    priv->scan_task =
+      *static_cast<std::shared_ptr<arrow::dataset::ScanTask> *>(g_value_get_pointer(value));
+    break;
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+gad_scan_task_get_property(GObject *object,
+                           guint prop_id,
+                           GValue *value,
+                           GParamSpec *pspec)
+{
+  switch (prop_id) {
+  default:
+    G_OBJECT_WARN_INVALID_PROPERTY_ID(object, prop_id, pspec);
+    break;
+  }
+}
+
+static void
+gad_scan_task_init(GADScanTask *object)
+{
+  auto priv = GAD_SCAN_TASK_GET_PRIVATE(object);
+  new(&priv->scan_task) std::shared_ptr<arrow::dataset::ScanTask>;
+}
+
+static void
+gad_scan_task_class_init(GADScanTaskClass *klass)
+{
+  GObjectClass *gobject_class;
+  GParamSpec *spec;
+
+  gobject_class = G_OBJECT_CLASS(klass);
+
+  gobject_class->finalize     = gad_scan_task_finalize;
+  gobject_class->set_property = gad_scan_task_set_property;
+  gobject_class->get_property = gad_scan_task_get_property;
+
+  spec = g_param_spec_pointer("scan-task",
+                              "ScanTask",
+                              "The raw std::shared<arrow::dataset::ScanTask> *",
+                              static_cast<GParamFlags>(G_PARAM_WRITABLE |
+                                                       G_PARAM_CONSTRUCT_ONLY));
+  g_object_class_install_property(gobject_class, PROP_SCAN_TASK, spec);
+}
+
+/**
+ * gad_scan_task_execute:
+ * @scan_task: A #GADScanTask.
+ * @error: (nullable): REturn location for a #GError or %NULL.
+ *
+ * Returns: (nullable): A newly created #GArrowRecordBatchIterator,
+ *   or %NULL on error.
+ *
+ * Since: 1.0.0
+ */
+GArrowRecordBatchIterator *gad_scan_task_execute(GADScanTask *scan_task,
+                                                 GError **error)
+{
+  auto priv = GAD_SCAN_TASK_GET_PRIVATE(scan_task);
+  auto arrow_result = priv->scan_task->Execute();
+  if (garrow::check(error, arrow_result, "[datasets][scan-task][execute]")) {
+    // FIXME: We cannot copy arrow::Iterator<T> here.
+    auto arrow_record_batch_iteraor = *arrow_result;
```

Comment:
@kou We cannot instantitate `GArrowRecordBatchIterator` here because `RecordBatchIterator` class is not copyable.  The reason of the uncopyability is due to its `std::unique_ptr<T>` member.  I'm considering how to avoid this constraint.  Do you have some ideas?

Reply:
I could resolve this by `std::move`.